File: backend/app/api/refine.py
"""优化层接口：对已经出好的那一段画面做二次处理（超分 / 插帧 / 重做）。

`plan` 与 `run` 分成两个端点（与 `adopt` / `sequence` 同一个习惯）：一次批量超分可能起十几
个任务、跑掉几十分钟显存，所以先给账单再动手。产出是**同一个镜头上的新版本**，所以这里
没有任何「采用」端点——那件事全工程只有 `POST /versions/{id}/current` 一个入口。
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel, Field

from app.persistence.models_gen import REFINE_KINDS
from app.services.refine import LABELS, refine

router = APIRouter(tags=["refine"])


class RefineBody(BaseModel):
    version_ids: list[str] | None = Field(
        default=None, description="就处理这几版（版本轨里勾的那些，最精确）"
    )
    shot_ids: list[str] | None = Field(
        default=None, description="这几个镜头**采用的那一版**；没采用过的会被跳过并说明"
    )
    scene_id: str | None = Field(default=None, description="整幕批量处理")
    kind: str = Field(default="upscale", description=" / ".join(REFINE_KINDS))
    preset: str | None = Field(
        default=None, description="临时换一份图；不传就是设置里的二次处理预设 → 默认预设"
    )
    prompt: str | None = Field(default=None, description="有些处理图也收 prompt，可留空")
    priority: int = 100
    extra: dict[str, Any] | None = None


@router.get("/refine/kinds")
async def kinds() -> list[dict[str, Any]]:
    """有哪几种二次处理。种类是后端的事实（`models_gen.REFINE_KINDS`），前端不要硬编码。"""
    return [{"kind": k, "label": LABELS.get(k, k)} for k in REFINE_KINDS]


@router.post("/projects/{pid}/refine/plan")
async def plan(pid: str, body: RefineBody) -> dict[str, Any]:
    """只出账单：处理哪几段、用哪份图、哪几个跳过为什么。**一个任务都不入队。**"""
    return await refine.plan(
        pid,
        version_ids=body.version_ids,
        shot_ids=body.shot_ids,
        scene_id=body.scene_id,
        kind=body.kind,
        preset=body.preset,
    )


@router.post("/projects/{pid}/refine/run", status_code=201)
async def run(pid: str, body: RefineBody) -> dict[str, Any]:
    """按账单入队。原来那一版一条都不动（只增不改），不满意就再采用回去。"""
    return await refine.run(
        pid,
        version_ids=body.version_ids,
        shot_ids=body.shot_ids,
        scene_id=body.scene_id,
        kind=body.kind,
        preset=body.preset,
        priority=body.priority,
        prompt=body.prompt,
        extra=body.extra,
    )


@router.get("/projects/{pid}/versions/{version_id}/lineage")
async def lineage(pid: str, version_id: str) -> dict[str, Any]:
    """这一版的谱系（原始 v1 → 超分 v2 → …）。版本轨靠它画出一条线而不是几条孤立版本。"""
    return await refine.lineage(pid, version_id)

File: frontend/src/shared/api/refine.ts
/**
 * 视频二次处理接口（Refine，超分 / 插帧 / 重做）。
 *
 * 对应后端 app/api/refine.py + services/refine.py：
 *   - kinds: 获取二次处理种类及说明；
 *   - plan: 优化账单；
 *   - run: 按账单入队，产出新版本并记录 parent_version_id；
 *   - lineage: 查询版本谱系链（祖先与衍生版本）。
 */

import { api } from './client'
import type { GenerationVersion, Job } from './generation'
import type { RouteBinds, RouteSource } from './projects'

export interface RefineKind {
  kind: 'upscale' | 'interpolate' | 'recut'
  label: string
}

export interface RefinePlanItem {
  version_id: string
  version_no: number
  shot_id: string
  shot_index_no: number
  scene_id: string
  kind: string
  duration: number
  asset_id: string | null
}

export interface RefinePlanSkipped {
  target: string
  error: {
    code: string
    title: string
    detail: string
    suggestions: string[]
  }
}

export interface RefinePlanResult {
  kind: string
  kind_label: string
  preset: string | null
  preset_ready: boolean
  preset_detail: string
  /**
   * 走哪条路。**界面照 `binds` 决定要不要显示「处理预设」那一行**：走 REST 时这个工程
   * 根本没有预设这回事，写「默认视频预设」是在说一个不存在的东西；绑定那条路做不了
   * 二次处理，那句话由 `preset_detail` + `how_to` 说清（硬约束 1、4）。
   */
  route: {
    provider: string
    label: string
    /** `project` 工程显式选了 / `settings` 跟随设置页 / `default` 谁都没选过。 */
    source: RouteSource
    /** `preset` / `base_url` / `workflow`，未知调用方式是空串。 */
    binds: RouteBinds
  }
  items: RefinePlanItem[]
  skipped: RefinePlanSkipped[]
  total: number
  blocked: boolean
  /** 这条路做不了时的出路（四要素里那几条建议）。**必须原样显示**（硬约束 4）。 */
  how_to: string[]
}

export interface RefineBody {
  version_ids?: string[]
  shot_ids?: string[]
  scene_id?: string
  kind?: 'upscale' | 'interpolate' | 'recut'
  preset?: string
  seed?: number
  priority?: number
}

export interface VersionLineageResult {
  version_id: string
  ancestors: GenerationVersion[]
  children: GenerationVersion[]
}

export const refineApi = {
  kinds: () => api.get<RefineKind[]>('/refine/kinds'),

  plan: (pid: string, body: RefineBody) =>
    api.post<RefinePlanResult>(`/projects/${pid}/refine/plan`, body),

  run: (pid: string, body: RefineBody) =>
    api.post<Job[]>(`/projects/${pid}/refine/run`, body),

  lineage: (pid: string, versionId: string) =>
    api.get<VersionLineageResult>(`/projects/${pid}/versions/${versionId}/lineage`),
}
